src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs

```diff
@@ -335,9 +336,7 @@ public static bool HasChanges(XDocument newFile, string path, ILogger log)
 
                 if (existing != null)
                 {
-                    // Use a simple string compare to check if the files match
-                    // This can be optimized in the future, but generally these are very small files.
-                    return !newFile.ToString().Equals(existing.ToString(), StringComparison.Ordinal);
+                    return !XDocument.DeepEquals(existing, newFile);
```

Comment:
I've read about problems with XDocument.DeepEquals in the past, where semantically equivalent XDocuments return false, can you point to the existing test coverage or add some more?

Reply:
> I've read about problems with XDocument.DeepEquals in the past, where semantically equivalent XDocuments return false

I agree and that's the reason "xmlns" attribute is explicitly added [here](https://github.com/NuGet/NuGet.Client/blob/c22dfaaecec56c7e664d90a374d960267bd0ac3c/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs#L188). I realized during debugging that in-memory `XDocument` has  `Root.FirstAttribute` and `Root.LastAttribute` set to `"ToolsVersion"`. On the other hand when `XDocument` is read from file system `Root.FirstAttribute` is set to `"ToolsVersion"` and `Root.LastAttribute` is set to `"xmlns"` attribute. `XDocument.DeepEquals` returned false because of this difference.

Reply:
> can you point to the existing test coverage or add some more?

Link to existing unit tests.

https://github.com/NuGet/NuGet.Client/blob/64f2febf7d1de8b0228eca5afed97a96c5a30bba/test/NuGet.Core.Tests/NuGet.Commands.Test/BuildAssetsUtilsTests.cs#L901

https://github.com/NuGet/NuGet.Client/blob/64f2febf7d1de8b0228eca5afed97a96c5a30bba/test/NuGet.Core.Tests/NuGet.Commands.Test/BuildAssetsUtilsTests.cs#L930

Link to a sample func test that covers this fix.

https://github.com/NuGet/NuGet.Client/blob/64f2febf7d1de8b0228eca5afed97a96c5a30bba/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs#L7514

I will be happy to add more coverage, if required.

Reply:
> Is this necessary?
It's already set in nuget.g.props/nuget.g.targets, how does that happen?

The `XNamespace` is set implicitly now because of [this overloaded operator](https://docs.microsoft.com/en-us/dotnet/api/system.xml.linq.xnamespace.op_addition?view=netcore-3.1) implementation. `XDocument.DeepEquals` API returned `false` because in-memory `XDocument` has  `Root.FirstAttribute` and `Root.LastAttribute` set to `"ToolsVersion"`. On the other hand when `XDocument` that is read from file system `Root.FirstAttribute` is set to `"ToolsVersion"` and `Root.LastAttribute` is set to `"xmlns"` attribute. Hence, I have added this `xmlns` attribute explicitly by referring [doc](https://docs.microsoft.com/en-us/dotnet/api/system.xml.linq.xnamespace?view=netcore-3.1#creating-a-default-namespace) so that in-memory `xml` tree matches with the one read from file system.
https://github.com/NuGet/NuGet.Client/blob/64f2febf7d1de8b0228eca5afed97a96c5a30bba/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs#L188